### ISSUE TITLE
Unable to resolve groovy source in test files

### ISSUE THREAD

**Issue by gprasann** [groovy/groovy-eclipse#321]
I am fairly new to Groovy. I have created a groovy project and have converted it to maven. I use Eclipse Kepler with groovy support for development.
When I run the maven build locally, everything works fine, but when I run the build on Jenkins, I see unresolved groovy source file errors and all the compilation errors are only in test cases.
Does anyone know why do I have an issue on Jenkins? Any pointers on this would be vey helpful.
<build> section in my pom.xml file looks like below. Thank you.

	<build>
		<sourceDirectory>src/groovy</sourceDirectory>
		<testSourceDirectory>test/groovy</testSourceDirectory>
		<plugins>
			<plugin>
				<artifactId>maven-compiler-plugin</artifactId>
				
				<version>3.1</version>
				<configuration>
					<compilerId>groovy-eclipse-compiler</compilerId>
					<source>${jdk.version}</source>
					<target>${jdk.version}</target>
				</configuration>
				<dependencies>
					<dependency>
						<groupId>org.codehaus.groovy</groupId>
						<artifactId>groovy-eclipse-compiler</artifactId>
						<version>2.9.1-01</version>
					</dependency>
					<dependency>
						<groupId>org.codehaus.groovy</groupId>
						<artifactId>groovy-eclipse-batch</artifactId>
						<version>2.3.7-01</version>
					</dependency>
				</dependencies>
				<executions>
					<execution>
						<phase>test-compile</phase>
						<goals>
							<goal>testCompile</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.grails</groupId>
				<artifactId>grails-maven-plugin</artifactId>
				<version>2.4.4</version>
				<configuration>
					<skipTests>true</skipTests>
					<fork>true</fork>
				</configuration>
				<extensions>true</extensions>
				<executions>
					<execution>
						<goals>
							<goal>package-plugin</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-surefire-plugin</artifactId>
				<version>2.9</version>
				<executions>
					<execution>
						<phase>test</phase>
						<goals>
							<goal>test</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>build-helper-maven-plugin</artifactId>
				<version>1.5</version>
				<executions>
					<execution>
						<id>add-source</id>
						<phase>generate-sources</phase>
						<goals>
							<goal>add-source</goal>
						</goals>
						<configuration>
							<sources>
								<source>grails-app/services</source>
							</sources>
						</configuration>
					</execution>
					<execution>
						<id>add-test-source</id>
						<phase>generate-test-sources</phase>
						<goals>
							<goal>add-test-source</goal>
						</goals>
						<configuration>
							<sources>
								<source>test/groovy</source>
								<source>test/java</source>
							</sources>
						</configuration>
					</execution>
					<execution>
						<id>add-test-resource</id>
						<phase>generate-test-sources</phase>
						<goals>
							<goal>add-test-resource</goal>
						</goals>
						<configuration>
							<resources>
								<resource>
									<directory>test/groovy</directory>
									<excludes>
										<exclude>**/*.groovy</exclude>
									</excludes>
								</resource>
							</resources>
						</configuration>
					</execution>
				</executions>
			</plugin>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>cobertura-maven-plugin</artifactId>
				<version>2.7</version>
				<configuration>
					<formats>
						<format>html</format>
						<format>xml</format>
					</formats>
				</configuration>
			</plugin>
		</plugins>
	</build>

**Comment by eric-milles** [comment]
I'm not 100% sure that the Groovy Maven plug-in allows overriding of the sourceDirectory and testSourceDirectory.  From what I have seen, it is hardcoded to src/main/groovy and src/test/groovy.  Could you at least try that out to be sure it is not the issue?

**Comment by gprasann** [comment]
Hi Eric,
Thank you for your response on this. I have tried to place the source under
src/main/java
src/main/groovy

and tests under
src/test/java
src/test/groovy

but still no luck. I see the same issue. Any other pointers on this? I am trying all the possible options :(

Thanks

**Comment by PascalSchumacher** [comment]
@gprasann Setting up sourcefolders for the groovy-eclipse-maven plugin is a bit complicated. Have your read https://github.com/groovy/groovy-eclipse/wiki/Groovy-Eclipse-Maven-plugin#setting-up-the-source-folders ?

**Comment by eric-milles** [comment]
When you build from Eclipse, are you just running the builder (Ctrl-B or Build Project/All)?  Or are you running Maven's Run As > Maven Build.

Did you start from the Groovy Maven<https://github.com/groovy/groovy-eclipse/wiki/Groovy-Eclipse-Maven-plugin> wiki page?  If there is something incorrect there, I'd like to get it fixed.

**Comment by gprasann** [comment]
Thank you @PascalSchumacher and @eric-milles for your replies.

@PascalSchumacher - I have read that link and tried all of the options suggested by moving the source/test files under src/main and directly under src/ with different build options.
For some reason it still fails.

@eric-milles - I am running it as Maven build and not the eclipse build. I have tried the options from the link provided, but no luck.

I am still trying to figureout the issue :(

Thanks

**Comment by gprasann** [comment]
Now I tried to add _maven-surefire-plugin_ as well and after which the grovy testcases are not at all compiled by the compiler. It really is behaving crazy. I tried to even add build-helper-maven-plugin by moving the source/test files outside of src/main and src/test, still no luck.

Thanks

**Comment by eric-milles** [comment]
I created a very simple Groovy-Maven project and could compile both main and test sources using Run As > Maven Build... with goals `compile` and `test-compile`

![groovy-maven](https://user-images.githubusercontent.com/18193802/28584239-a01c68b8-7131-11e7-9fd0-fb9a75ae5a41.png)

```xml
<project
 xmlns="http://maven.apache.org/POM/4.0.0"
 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
                     http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <artifactId>Mvn</artifactId>
  <groupId>a.b.c</groupId>
  <version>0.0.1-SNAPSHOT</version>

  <build>
    <plugins>
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.6.1</version>
        <configuration>
          <compilerId>groovy-eclipse-compiler</compilerId>
          <source>1.8</source>
          <target>1.8</target>
        </configuration>
        <dependencies>
          <dependency>
            <groupId>org.codehaus.groovy</groupId>
            <artifactId>groovy-eclipse-compiler</artifactId>
            <version>2.9.2-01</version>
          </dependency>
          <dependency>
            <groupId>org.codehaus.groovy</groupId>
            <artifactId>groovy-eclipse-batch</artifactId>
            <version>2.4.3-01</version>
          </dependency>
        </dependencies>
      </plugin>
    </plugins>
  </build>

  <dependencies>
    <dependency>
      <groupId>org.codehaus.groovy</groupId>
      <artifactId>groovy-all</artifactId>
      <version>2.4.12</version>
    </dependency>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.12</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
</project>
```

When I run Maven > Update Project... it added the Groovy nature and the source folders src/main/groovy and src/test/groovy.  I needed files in src/main/java and src/test/java for this, as described in https://github.com/groovy/groovy-eclipse/wiki/Groovy-Eclipse-Maven-plugin#setting-up-the-source-folders

**Comment by gprasann** [comment]
Hi @eric-milles - Thank you very mmuch for your help on this.
I am not sure what the problem was with whatever I was doing.
Earlier I was creating a grails project and converting it to Maven and it was not working.

Now I followed your approach of creating the Maven project and then converting to Grails and it looks to be working fine.

I am unable to understand the differences between the two approaches, but I am happy that I am able to run it fine now :)

Thanks again all for your help on this.

**Comment by gprasann** [comment]
Hi @eric-milles - I got the exact issue now.
It is working fine when I use maven packaging **Jar**, but the moment I convert the packaging to **Pom**, the testcases wouldn't run along with maven build.
And when I use **compiler:testCompile** option for maven build, testcases wouldn't find the source files.

Any idea what I am missing here?

Thanks,
Prasann

**Comment by eric-milles** [comment]
@gprasann Can you describe the change I would make to my pom.xml to get to this state?  I've not used Maven a whole lot.

**Comment by gprasann** [comment]
Hi @eric-milles, Thank you for the reply. You need to add <packaging> tag to your pom.xml file as below

<packaging>pom</packaging>

With this we would be able to build grails plugin projects as zip artifacts instead of jar.

Thanks,
Prasann

**Comment by eric-milles** [comment]
So once changed to a pom project, I think Maven is expecting a module that contains the sources and tests to be packaged as jar or war or whatever.  With pom packaging, my build does not even compile sources.

**Comment by gprasann** [comment]
Hi @eric-milles - Looks like I have craked it. I am using the default packaging (jar) and I am using the grails-maven-plugin with goal as package-plugin.

With this I had to move source files out of src/main directly under src. And same for test files. Otherwise even the test files are packaged in the plugin artifact (not really sure why)

I also had to use build-helper-maven-plugin to add source and test folders.

Now after doing all this, I am able to build the grails plugin fine with maven.

Hopefully this will help someone who has same problem :)

And Thank you again for your help.
